Snowflake.Data/Core/Authenticator/ExternalBrowserAuthenticator.cs

```diff
@@ -0,0 +1,243 @@
+
+                logger.Debug("Get IdpUrl and ProofKey");
+                var authenticatorRestRequest = BuildAuthenticatorRestRequest(localPort);
+                var authenticatorRestResponse = session.restRequester.Post<AuthenticatorResponse>(authenticatorRestRequest);
+                authenticatorRestResponse.FilterFailedResponse();
+
+                var idpUrl = authenticatorRestResponse.data.ssoUrl;
+                proofKey = authenticatorRestResponse.data.proofKey;
+
+                logger.Debug("Open browser");
+                StartBrowser(idpUrl);
+
+                logger.Debug("Get the redirect SAML request");
+                var context = httpListener.GetContext();
+                var request = context.Request;
+                samlResponseToken = ValidateAndExtractToken(request);
+                HttpListenerResponse response = context.Response;
+                try
+                {
+                    using (var output = response.OutputStream)
+                    {
+                        output.Write(SUCCESS_RESPONSE, 0, SUCCESS_RESPONSE.Length);
+                    }
+                }
+                catch (Exception e)
+                {
+                    // Ignore the exception as it does not affect the overall authentication flow
+                }
+
+                httpListener.Stop();
+            }
+
+            logger.Debug("Send login request");
+            var loginResponse = session.restRequester.Post<LoginResponse>(
+                BuildExternalBrowserLoginRequest(samlResponseToken, proofKey)
+                );
+            session.ProcessLoginResponse(loginResponse);
+        }
+
+        private static int GetRandomUnusedPort()
+        {
+            var listener = new TcpListener(IPAddress.Loopback, 0);
+            listener.Start();
+            var port = ((IPEndPoint)listener.LocalEndpoint).Port;
+            listener.Stop();
+            return port;
+        }
+
+        private static HttpListener GetHttpListener(int port)
+        {
+            string redirectURI = string.Format("http://{0}:{1}/", IPAddress.Loopback, port);
+            HttpListener listener = new HttpListener();
+            listener.Prefixes.Add(redirectURI);
+            return listener;
+        }
+
+        private static void StartBrowser(string url)
+        {
+            // The following code is learnt from https://brockallen.com/2016/09/24/process-start-for-urls-on-net-core/
+#if NET46
+            // .net standard would pass here
+            Process.Start(url);
+#else
+            // hack because of this: https://github.com/dotnet/corefx/issues/10361
+            if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                url = url.Replace("&", "^&");
+                Process.Start(new ProcessStartInfo("cmd", $"/c start {url}") { CreateNoWindow = true });
+            }
+            else if (RuntimeInformation.IsOSPlatform(OSPlatform.Linux))
```

Comment:
I don't think we support .NET on Linux as of today. Unless we test it, remove this branch.

Reply:
Do we support it on Mac?